xonsh/cli_utils.py

```diff
@@ -51,58 +94,289 @@ def get_doc(func: tp.Callable, parameter: str = None):
+            )
+        return text
+
+    def _format_action_invocation(self, action):
+        from pygments.token import Name
+
+        text = super()._format_action_invocation(action)
+        return self.colorize((Name.Attribute, text))
+
+
+def get_argparse_formatter_class():
+    from xonsh.built_ins import XSH
+    from xonsh.platform import HAS_PYGMENTS
+
+    if (
+        hasattr(sys, "stderr")
+        and sys.stderr.isatty()
+        and XSH.env.get("XONSH_INTERACTIVE")
+        and HAS_PYGMENTS
+    ):
+        return RstHelpFormatter
+    return ap.RawTextHelpFormatter
+
+
+class ArgParser(ap.ArgumentParser):
+    """Sub-class of ArgumentParser with special methods to nest commands"""
+
+    def __init__(self, **kwargs):
+        if "formatter_class" not in kwargs:
+            kwargs["formatter_class"] = get_argparse_formatter_class()
+
+        super().__init__(**kwargs)
+        self.commands = None
+
+    def add_command(
+        self, func: tp.Callable, args: tp.Optional[tp.Iterable[str]] = None, **kwargs
+    ):
+        """
+            create a sub-parser and call this function during dispatch
+
+        Parameters
+        ----------
+        func
+            a type-annotated function that will be used to create ArgumentParser instance.
+            All parameters that start with ``_`` will not be added to parser arguments.
+            Use _stdout, _stack ... to receive them from callable-alias/commands.
+            Use _parser to get the generated parser instance.
+            Use _args to get what is passed from sys.argv
+            Use _parsed to get result of ``parser.parse_args``
+        args
+            if given only add these arguments to the parser.
+            Otherwise all parameters to the function without `_` prefixed
+            in their name gets added to the parser.
+        kwargs
+            passed to ``subparser.add_parser`` call
+
+        Returns
+        -------
+            result from ``subparser.add_parser``
+        """
+        if not self.commands:
+            self.commands = self.add_subparsers(title="commands", dest="command")
+
+        doc = get_doc(func)
+        kwargs.setdefault("description", doc)
+        kwargs.setdefault("help", doc)
+        parser = self.commands.add_parser(kwargs.pop("prog", func.__name__), **kwargs)
+        add_args(parser, func, allowed_params=args)
         return parser
 
 
-def dispatch(**ns):
+def dispatch(parser: ap.ArgumentParser, args=None, **ns):
     """call the sub-command selected by user"""
     import inspect
 
+    parsed = parser.parse_args(args)
+    ns["_parsed"] = parsed
+    ns.update(vars(parsed))
+
     func = ns[_FUNC_NAME]
     sign = inspect.signature(func)
     kwargs = {}
-    for name, _ in sign.parameters.items():
-        kwargs[name] = ns[name]
+    for name, param in sign.parameters.items():
+        default = None
+        if param.default != inspect.Parameter.empty:
+            default = param.default
```

Comment:
Isn't this already done in `add_args`?

Reply:
yes you are right.

Reply:
I see, can you add a comment explaining this?

Reply:
ping :)

Reply:
ping :)

Reply:
done https://github.com/xonsh/xonsh/pull/4391/commits/7ca1cf2d5f0d97cb546b4c21d2d2cb8cf178db50